File: frontend/src/types/tracking.ts
import type { LetterStatus } from "./letter"

export interface TrackingEvent {
  event_type: string
  actor_type: string
  created_at: string
}

export interface TrackingRead {
  reference: string
  status: LetterStatus
  subject: string
  sender_first_name: string
  sender_last_name: string
  recipient_first_name: string
  recipient_last_name: string
  created_at: string
  events: TrackingEvent[]
}

export interface AccessLetterView {
  reference: string
  sender_first_name: string
  sender_last_name: string
  subject: string
  message: string | null
  status: LetterStatus
  has_document: boolean
  acknowledgment_of_receipt: boolean
  created_at: string
}

File: backend/app/schemas/tracking.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel

from app.models.enums import ActorType, LetterEventType, LetterStatus


class TrackingEvent(BaseModel):
    event_type: LetterEventType
    actor_type: ActorType
    created_at: datetime

    model_config = {"from_attributes": True}


class TrackingRead(BaseModel):
    reference: str
    status: LetterStatus
    subject: str
    sender_first_name: str
    sender_last_name: str
    recipient_first_name: str
    recipient_last_name: str
    created_at: datetime
    events: list[TrackingEvent]


class AccessLetterView(BaseModel):
    reference: str
    sender_first_name: str
    sender_last_name: str
    subject: str
    message: str | None
    status: LetterStatus
    has_document: bool
    acknowledgment_of_receipt: bool
    created_at: datetime
